packages/ws/src/ws/WebSocketShard.ts

```diff
@@ -0,0 +1,515 @@
+		const controller = new AbortController();
+		const timeout = timeoutDuration ? setTimeout(() => controller.abort(), timeoutDuration).unref() : null;
+		await once(this, event, { signal: controller.signal });
+		if (timeout) {
+			clearTimeout(timeout);
+		}
+	}
+
+	public async send(payload: GatewaySendPayload) {
+		if (!this.connection) {
+			throw new Error("WebSocketShard wasn't connected");
+		}
+
+		if (this.status !== WebSocketShardStatus.Ready && !ImportantGatewayOpcodes.has(payload.op)) {
+			throw new Error('Tried to send a message before the shard was ready');
+		}
+
+		await this.sendQueue.wait();
+
+		if (--this.sendRateLimitState.remaining <= 0) {
+			if (this.sendRateLimitState.resetAt < Date.now()) {
+				await sleep(Date.now() - this.sendRateLimitState.resetAt);
+			}
+
+			this.sendRateLimitState = {
+				remaining: 119,
+				resetAt: Date.now() + 60_000,
+			};
+		}
+
+		this.sendQueue.shift();
+		this.connection.send(JSON.stringify(payload));
+	}
+
+	private async identify() {
+		this.debug(['Identifying']);
+		const d: GatewayIdentifyData = {
+			token: this.strategy.options.token,
+			properties: this.strategy.options.identifyProperties,
+			intents: this.strategy.options.intents,
+			compress: this.useIdentifyCompress,
+			shard: [this.id, this.strategy.options.shardCount],
+		};
+
+		if (this.strategy.options.largeThreshold) {
+			d.large_threshold = this.strategy.options.largeThreshold;
+		}
+
+		if (this.strategy.options.initialPresence) {
+			d.presence = this.strategy.options.initialPresence;
+		}
+
+		await this.send({
+			op: GatewayOpcodes.Identify,
+			d,
+		});
+
+		await this.waitForEvent(WebSocketShardEvents.Ready, this.strategy.options.readyTimeout);
+		this.status = WebSocketShardStatus.Ready;
+	}
+
+	private resume(session: SessionInfo) {
+		this.debug(['Resuming session']);
+		this.status = WebSocketShardStatus.Resuming;
+		this.replayedEvents = 0;
+		return this.send({
+			op: GatewayOpcodes.Resume,
+			d: {
+				token: this.strategy.options.token,
+				seq: session.sequence,
+				session_id: session.sessionId,
+			},
+		});
+	}
+
+	private async heartbeat(requested = false) {
+		if (!this.isAck && !requested) {
+			return this.destroy({ reason: 'Zombie connection', recover: WebSocketShardDestroyRecovery.Resume });
+		}
+
+		await this.send({
+			op: GatewayOpcodes.Heartbeat,
+			d: this.session?.sequence ?? null,
+		});
+
+		this.lastHeartbeatAt = Date.now();
+		this.isAck = false;
+	}
+
+	private async unpackMessage(data: Buffer | ArrayBuffer, isBinary: boolean): Promise<GatewayReceivePayload | null> {
+		const decompressable = new Uint8Array(data);
+
+		// Deal with no compression
+		if (!isBinary) {
+			return JSON.parse(this.textDecoder.decode(decompressable)) as GatewayReceivePayload;
+		}
+
+		// Deal with identify compress
+		if (this.useIdentifyCompress) {
+			return new Promise((resolve, reject) => {
```

Comment:
It's not always zlib'd. You need to detect it 🤔

Reply:
unsure what you mean? this exits if it's not zlibed:
https://github.com/discordjs/discord.js/blob/71117182ac8fe2b7de1b34d6b58a32fb6a270d6d/packages/ws/src/ws/WebSocketShard.ts#L277-L280

Reply:
Is isBinary set *only* when the data is not a string? How does it know?

Reply:
I honestly have no clue, but from my testing identifyCompress worked

Reply:
Alrighty